File: frontend/src/types/gameState.ts
export type TurnPhase = 
  | 'untap' 
  | 'upkeep' 
  | 'draw' 
  | 'main1' 
  | 'combat_start' 
  | 'combat_attack' 
  | 'combat_block' 
  | 'combat_damage' 
  | 'combat_end' 
  | 'main2' 
  | 'end' 
  | 'cleanup';

export type CardZone = 
  | 'library' 
  | 'hand' 
  | 'battlefield' 
  | 'graveyard' 
  | 'exile' 
  | 'commander';

export interface GameCardImageUris {
  small?: string;
  normal?: string;
  large?: string;
  art_crop?: string;
  border_crop?: string;
}

export interface GameCardFace {
  name?: string;
  mana_cost?: string;
  type_line?: string;
  oracle_text?: string;
  power?: string;
  toughness?: string;
  image_uris?: GameCardImageUris;
}

export interface GameCard {
  id: number;
  card_scryfall_id: string;
  card_name: string;
  mana_cost?: string;
  cmc?: number;
  type_line?: string;
  oracle_text?: string;
  colors?: string[];
  power?: string;
  toughness?: string;
  keywords?: string[];
  image_uris?: GameCardImageUris;
  card_faces?: GameCardFace[];
  zone: CardZone;
  position: number;
  is_tapped: boolean;
  is_face_up: boolean;
  battlefield_x?: number;
  battlefield_y?: number;
  is_attacking: boolean;
  is_blocking: boolean;
  damage_received: number;
  played_as_side?: number;
}

export interface GameCardInBattlefield extends Omit<GameCard, 'zone' | 'damage_received'> {
  battlefield_x?: number;
  battlefield_y?: number;
  position: number;
}

export interface PlayerGameState {
  id: number;
  user_id: number;
  username: string;
  player_order: number;
  is_active: boolean;
  life_total: number;
  poison_counters: number;
  library: GameCard[];
  hand: GameCard[];
  battlefield: GameCardInBattlefield[];
  graveyard: GameCard[];
  exile: GameCard[];
  commander: GameCard[];
  mana_pool: Record<string, number>;
}

export interface GameState {
  id: number;
  game_room_id: number;
  current_turn: number;
  active_player_id: number;
  active_player_username: string;
  current_phase: TurnPhase;
  starting_player_id: number;
  players: PlayerGameState[];
  created_at: string;
  logs?: GameLog[];
  game_mode?: string;
}

export const TURN_PHASE_LABELS: Record<TurnPhase, string> = {
  untap: 'Untap',
  upkeep: 'Upkeep',
  draw: 'Draw',
  main1: 'Main Phase',
  combat_start: 'Combat Start',
  combat_attack: 'Attack',
  combat_block: 'Block',
  combat_damage: 'Combat Damage',
  combat_end: 'Combat End',
  main2: 'Main Phase 2',
  end: 'End',
  cleanup: 'Cleanup',
};

export interface GameLog {
  id: number;
  game_id: number;
  player_id: number;
  action_type: string;
  card_id?: number;
  card_name?: string;
  from_zone?: string;
  to_zone?: string;
  message: string;
  created_at: string;
}

export interface CardSideOption {
  side_index: number;
  name: string;
  mana_cost?: string;
  type_line?: string;
  image_url?: string;
}

export interface ChooseCardSideResponse {
  card_id: number;
  card_name: string;
  sides: CardSideOption[];
  requires_side_selection: boolean;
}

export interface ValidPlayCard {
  card_id: number;
  card_name: string;
  zone: CardZone;
  mana_cost?: string;
  can_afford_mana: boolean;
  needs_side_selection: boolean;
  sides?: CardSideOption[];
}

export interface ValidPlaysResponse {
  current_phase: TurnPhase;
  can_cast_spells: boolean;
  can_play_land: boolean;
  available_mana: Record<string, number>;
  untapped_lands_count: number;
  plays: ValidPlayCard[];
}

File: backend/app/schemas/game_state.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List, Dict
from datetime import datetime
from app.models.game import TurnPhase, CardZone


class GameLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    game_id: int
    player_id: int
    action_type: str
    card_id: Optional[int] = None
    card_name: Optional[str] = None
    from_zone: Optional[str] = None
    to_zone: Optional[str] = None
    message: str
    created_at: datetime


class GameCardResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    card_scryfall_id: str
    card_name: str
    mana_cost: Optional[str] = None
    cmc: Optional[float] = None
    type_line: Optional[str] = None
    oracle_text: Optional[str] = None
    colors: Optional[list] = None
    power: Optional[str] = None
    toughness: Optional[str] = None
    keywords: Optional[list] = None
    image_uris: Optional[dict] = None
    card_faces: Optional[list] = None
    zone: CardZone
    position: int
    is_tapped: bool
    is_face_up: bool
    battlefield_x: Optional[float] = None
    battlefield_y: Optional[float] = None
    is_attacking: bool
    is_blocking: bool
    damage_received: int


class GameCardInBattlefieldResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    card_scryfall_id: str
    card_name: str
    mana_cost: Optional[str] = None
    type_line: Optional[str] = None
    oracle_text: Optional[str] = None
    power: Optional[str] = None
    toughness: Optional[str] = None
    image_uris: Optional[dict] = None
    card_faces: Optional[list] = None
    is_tapped: bool
    is_face_up: bool
    battlefield_x: Optional[float] = None
    battlefield_y: Optional[float] = None
    is_attacking: bool
    is_blocking: bool
    position: int


class PlayerGameStateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    username: str
    player_order: int
    is_active: bool
    life_total: int
    poison_counters: int
    mana_pool: Optional[Dict[str, int]] = {}
    library: List[GameCardResponse]
    hand: List[GameCardResponse]
    battlefield: List[GameCardInBattlefieldResponse]
    graveyard: List[GameCardResponse]
    exile: List[GameCardResponse]
    commander: List[GameCardResponse]


class GameStateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    game_room_id: int
    current_turn: int
    active_player_id: int
    active_player_username: str
    current_phase: TurnPhase
    starting_player_id: int
    players: List[PlayerGameStateResponse]
    created_at: datetime
    logs: Optional[List[GameLogResponse]] = []
    game_mode: Optional[str] = None


class DrawCardRequest(BaseModel):
    card_id: Optional[int] = None


class PlayCardRequest(BaseModel):
    card_id: int
    target_zone: CardZone = CardZone.BATTLEFIELD
    position: int = 0
    battlefield_x: Optional[float] = None
    battlefield_y: Optional[float] = None
    side_index: Optional[int] = None


class MoveCardRequest(BaseModel):
    card_id: int
    target_zone: CardZone
    position: int = 0


class MoveCardsRequest(BaseModel):
    cards: List[MoveCardRequest]


class TapCardRequest(BaseModel):
    card_id: int


class BattlefieldPositionRequest(BaseModel):
    card_id: int
    x: float
    y: float


class AdjustLifeRequest(BaseModel):
    amount: int


class AddManaRequest(BaseModel):
    color: str
    amount: int = 1


class TapLandForManaRequest(BaseModel):
    """Request to tap a land and add mana to the pool."""
    color: Optional[str] = None  # Required for hybrid lands, optional for single color lands


class GameActionErrorResponse(BaseModel):
    error_type: str
    message: str
    code: str


class CardSideOption(BaseModel):
    side_index: int
    name: str
    mana_cost: Optional[str] = None
    type_line: Optional[str] = None
    image_url: Optional[str] = None


class ChooseCardSideResponse(BaseModel):
    card_id: int
    card_name: str
    sides: List[CardSideOption]
    requires_side_selection: bool = True


class ValidPlayCard(BaseModel):
    card_id: int
    card_name: str
    zone: CardZone
    mana_cost: Optional[str] = None
    can_afford_mana: bool
    needs_side_selection: bool = False
    sides: Optional[List[CardSideOption]] = None


class ValidPlaysResponse(BaseModel):
    current_phase: TurnPhase
    can_cast_spells: bool
    can_play_land: bool
    available_mana: Dict[str, int]
    untapped_lands_count: int
    plays: List[ValidPlayCard]
